discovery/uyuni/uyuni.go

```diff
@@ -0,0 +1,397 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package uyuni
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolo/xmlrpc"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+const (
+	monitoringEntitlementLabel    = "monitoring_entitled"
+	prometheusExporterFormulaName = "prometheus-exporters"
+	uyuniXMLRPCAPIPath            = "/rpc/api"
+)
+
+// DefaultSDConfig is the default Uyuni SD configuration.
+var DefaultSDConfig = SDConfig{
+	RefreshInterval: model.Duration(1 * time.Minute),
+}
+
+// Regular expression to extract port from formula data
+var monFormulaRegex = regexp.MustCompile(`--(?:telemetry\.address|web\.listen-address)=\":([0-9]*)\"`)
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+}
+
+// SDConfig is the configuration for Uyuni based service discovery.
+type SDConfig struct {
+	Host            string         `yaml:"host"`
+	User            string         `yaml:"username"`
+	Pass            string         `yaml:"password"`
+	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
+}
+
+// Uyuni API Response structures
+type systemGroupID struct {
+	GroupID   int    `xmlrpc:"id"`
+	GroupName string `xmlrpc:"name"`
+}
+
+type networkInfo struct {
+	SystemID int    `xmlrpc:"system_id"`
+	Hostname string `xmlrpc:"hostname"`
+	IP       string `xmlrpc:"ip"`
+}
+
+type exporterConfig struct {
+	Address string `xmlrpc:"address"`
+	Args    string `xmlrpc:"args"`
+	Enabled bool   `xmlrpc:"enabled"`
+}
+
+type proxiedExporterConfig struct {
+	ProxyIsEnabled   bool           `xmlrpc:"proxy_enabled"`
+	ProxyPort        float32        `xmlrpc:"proxy_port"`
+	NodeExporter     exporterConfig `xmlrpc:"node_exporter"`
+	ApacheExporter   exporterConfig `xmlrpc:"apache_exporter"`
+	PostgresExporter exporterConfig `xmlrpc:"postgres_exporter"`
```

Comment:
Does it mean that every exporter would need a new prometheus release?

Reply:
Replaced it with extendable list of exporter configurations. Also, formulas can now be set in configuration.

Reply:
This shouldn't be required. They should be discovered, without the user having to configure them on the Prometheus side.

Reply:
I wasn't precise. They are not configured on Prometheus side now. The only information we provide in configuration is the list of formula names. `ExporterConfigs` are discovered from the pillar data.

Reply:
Why do you even need to provide formula names? An SDby default should just dump out everything, and leave it up to relabelling.

Reply:
Uyuni uses formulas for many different purposes not related with monitoring. Querying them all would not be efficient and useful. We could think of adding some logic in Uyuni to filter monitoring related formulas only if that's preferred.

Reply:
The ideal with Prometheus is that you're scraping applications directly, exporters only exist for when that's not possible. Accordingly every single deployed application is of interest, as all applications are interesting to monitoring - even if it's only for blackbox probing.

Reply:
Good point. Although we currently don't monitor applications directly it's a very good feature request.
I've added information about all installed formulas in metadata, so that this information can be used in relabeling configuration.

Reply:
Why is there still a formula config field?

Reply:
For formulas including exporters' configuration we need additional details from the pillar data. That are the ones I'm listing explicitly in configuration. Pulling pillar data for all formulas on the other hand would be an overkill in my opinion.

Reply:
My same comment still applies, all this metadata should be there for everything without having to provide a list of your services. Maybe provide a single endpoint that can provide all of this efficiently?